src/orange/orangeinp/UnitProto.cc

```diff
@@ -0,0 +1,180 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file orange/orangeinp/UnitProto.cc
+//---------------------------------------------------------------------------//
+#include "UnitProto.hh"
+
+#include <algorithm>
+
+#include "CsgObject.hh"
+#include "CsgTree.hh"
+#include "CsgTreeUtils.hh"
+#include "Transformed.hh"
+
+#include "detail/CsgUnit.hh"
+#include "detail/CsgUnitBuilder.hh"
+#include "detail/VolumeBuilder.hh"
+
+namespace celeritas
+{
+namespace orangeinp
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with required input data.
+ */
+UnitProto::UnitProto(Input&& inp) : input_{std::move(inp)}
+{
+    CELER_VALIDATE(input_, << "no fill, daughters, or volumes are defined");
+    CELER_VALIDATE(std::all_of(input_.materials.begin(),
+                               input_.materials.begin(),
+                               [](MaterialInput const& m) {
+                                   return static_cast<bool>(m);
+                               }),
+                   << "incomplete material definition(s)");
+    CELER_VALIDATE(std::all_of(input_.daughters.begin(),
+                               input_.daughters.begin(),
+                               [](DaughterInput const& d) {
+                                   return static_cast<bool>(d);
+                               }),
+                   << "incomplete daughter definition(s)");
+    CELER_VALIDATE(input_.boundary.zorder == ZOrder::media
+                       || input_.boundary.zorder == ZOrder::exterior,
+                   << "invalid exterior zorder '"
+                   << to_cstring(input_.boundary.zorder) << "'");
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Short unique name of this object.
+ */
+std::string_view UnitProto::label() const
+{
+    return input_.label;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get the boundary of this universe as an object.
+ */
+auto UnitProto::interior() const -> SPConstObject
+{
+    return input_.boundary.interior;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Get a list of all daughter protos.
+ */
+auto UnitProto::daughters() const -> VecProto
```

Comment:
Not sure if this is enough work to cross into "calc_daughters"

Reply:
Maybe... maybe `get_daughters`? I'll revisit this.